Ensure ordering of offset commit requests

prestona opened the pull request:
Add locking to prevent concurrent calls to commit from potentially being re-ordered between determining which offsets to include in the offset commit request, and sending the request to Kafka.

Note: we think this can be improved upon (by not adding more locks). A potential improvement would be to use the [broker lock](https://github.com/IBM/sarama/blob/d2246cc9b9df113f3ffed7c442fd81e8b9cf736e/broker.go#L31), and only hold this until the request is sent (avoiding holding a lock for the entire network round-trip time). However, it looks like #2409 has made [sendAndReceive](https://github.com/IBM/sarama/blob/d2246cc9b9df113f3ffed7c442fd81e8b9cf736e/broker.go#L1038) hold the broker lock until a response is received from the broker, so we need to understand that a little better first.

Fixes: #2940

dnwe reviewed (approved):
Fix LGTM, though it would be helpful if we could possibly get a reproduce of the out-of-order commits into the FV suite in a followup PR?

prestona commented:
> Fix LGTM, though it would be helpful if we could possibly get a reproduce of the out-of-order commits into the FV suite in a followup PR?

Yep. Good point. The manual reproduce needed quite a few repetitions before we hit the window, but we can take a look and see what an FV suite test would look like, and how often it can be made to detect the condition.

puellanivis commented on offset_manager_test.go:
```diff
@@ -105,6 +107,111 @@ func TestNewOffsetManager(t *testing.T) {
+					for partition, offset := range req.blocks["topic"] {
+						last := lastOffset[partition]
+						if last > offset.offset {
+							outOfOrder =
+								fmt.Sprintf("out of order commit to partition %d, current committed offset: %d, offset in request: %d",
+									partition, last, offset.offset)
```
Consider using a `sync/atomic.Value` with a pointer to a `struct{ partition, last, offset IntType }` shape struct, to avoid the inherent serialization caused by mutex locking.

prestona replied:
Good idea. I ended up using an `atomic.Pointer, as it avoids the need to make a type assertion when calling `Load()`.

puellanivis commented on offset_manager_test.go:
```diff
@@ -105,6 +107,111 @@ func TestNewOffsetManager(t *testing.T) {
+	const numPartitions = 10
+	const commitsPerPartition = 1000
+
+	wg := &sync.WaitGroup{}
```
Prefer simply declaring a variable `var wg sync.WaitGroup` rather than assigning a zero value / pointer to zero value.

prestona replied:
Fixed.

puellanivis commented on offset_manager_test.go:
```diff
@@ -105,6 +107,111 @@ func TestNewOffsetManager(t *testing.T) {
+	defer safeClose(t, testClient)
+	om, err := NewOffsetManagerFromClient("group", testClient)
+	if err != nil {
+		t.Error(err)
```
Can the test keep going after this?  `om` is almost certainly a nil pointer, so the first use is going to `nil` pointer exception, right?

prestona replied:
Well spotted. Substituted with `t.Fatal(...)`.

puellanivis commented on offset_manager_test.go:
```diff
@@ -105,6 +107,111 @@ func TestNewOffsetManager(t *testing.T) {
+	}
+
+	wg.Wait()
+	mu.Lock()
+	defer mu.Unlock()
```
This lock should be unnecessary. The `wg.Wait()` ensures all writes should be complete, and performs all the synchronization necessary.

prestona replied:
Superseded by switch to sync/atomic.Pointer. The intent was to ensure that the error message assigned to `outOfOrder` (by the goroutine running the handler in the mock broker) was guaranteed to be observed by the goroutine running the test function.

puellanivis replied:
Yeah, I figured it was added out of an abundance of caution. The information here (helpful apart from even this code review) is that the `wg.Wait()` itself is synchronizing, as all calls to `wg.Done()` _happen before_ the return of `wg.Wait()`, the same as if you had used a mutex.

puellanivis commented on offset_manager_test.go:
```diff
@@ -105,6 +107,111 @@ func TestNewOffsetManager(t *testing.T) {
+// Test that the correct sequence of offset commit messages is sent to a broker when
+// multiple goroutines for a group are committing offsets at the same time
+func TestOffsetManagerCommitSequence(t *testing.T) {
+	mu := &sync.Mutex{}
```
Same as below, just go ahead and declare the variable instead: `var mu sync.Mutex`

prestona replied:
Superseded by switch to sync/atomic.Pointer - which I've declared as per your suggestion.

prestona commented on offset_manager_test.go:
```diff
@@ -105,6 +107,111 @@ func TestNewOffsetManager(t *testing.T) {
+			func() {
+				mu.Lock()
+				defer mu.Unlock()
+				if outOfOrder == "" {
```
Superseded by switch to sync/atomic.Pointer

prestona commented:
@puellanivis, thanks for taking the time to review. I hope I've addressed your concerns in 21e6e58. I've also cherry-picked this change onto #2947 to keep that consistent.

puellanivis reviewed (commented):
Looks acceptable.

dnwe commented:
Superseded by 2947
